### ISSUE TITLE
AR transformer loss function

### ISSUE THREAD

**Issue by lawrence-cj** [TransDiff/TransDiff#3]
Hi team, congrats on the release. I noticed the loss for AR transformer is set to False: https://github.com/TransDiff/TransDiff/blob/acaa08606f0ca35350acfddaf5a213edb2576325/models/transdiff.py#L183

I'm curious when will the loss function eq(2) in the paper be used? Thanks in advance for your answers.

<img width="1135" alt="Image" src="https://github.com/user-attachments/assets/b5f2151b-57d4-41c2-9c88-f54f1b04c29c" />

**Comment by TransDiff** [comment]
Hi. Thanks for your interest and patience. 
Firstly, it needs to be clarified that the param 'use_diversity_loss'  in function 'self.forward_encoder' is uesd to control whether use diversity loss.

https://github.com/TransDiff/TransDiff/blob/4a501e0c3046a3c44f571d9fad541405c0ffcf25/models/transdiff.py#L127C5-L127C77

Diversity loss is rarely used, only when the metric of diversity increases significantly in a certain epoch. So this part does not appear in the paper. And the metric of diversity  you can find in chapter 'Metric for Evaluating the Diversity of Image Semantic Features', which is defined as below:

<img width="480" alt="Image" src="https://github.com/user-attachments/assets/3a5bca7a-ee7d-42ad-823a-b9c73c6eea0b" />

Secondly, the loss function eq(2) in the paper is suitable for all AR Transformer models.  In TransDiff, the loss function is specifically defined as eq(8) in the paper:

<img width="480" alt="Image" src="https://github.com/user-attachments/assets/75225838-7742-45fa-8be4-e5f7671e0021" />

Specifically: the 'z' in the code represents the output of AR Transformer() in eq(8) and  the final loss function is calculated in
[[https://github.com/TransDiff/TransDiff/blob/main/models/transdiff.py#L165[
](https://github.com/TransDiff/TransDiff/blob/4a501e0c3046a3c44f571d9fad541405c0ffcf25/models/transdiff.py#L165C5-L165C39)](https://github.com/TransDiff/TransDiff/blob/4a501e0c3046a3c44f571d9fad541405c0ffcf25/models/transdiff.py#L165)](https://github.com/TransDiff/TransDiff/blob/4a501e0c3046a3c44f571d9fad541405c0ffcf25/models/transdiff.py#L165)

**Comment by lawrence-cj** [comment]
Thanks for your kind answering. 

If I understand right, the eq(2) loss is not specifically designed for TransDiff. The real loss we used here is and end to end diffusion loss? 

Besides, ‘x_n denotes the n-th known input latent’ means n-th latent feature which is pixel or latent level shaped as (h’,w’,c), not a token level shaped (1,1,c), no?

**Comment by TransDiff** [comment]
Q: If I understand right, the eq(2) loss is not specifically designed for TransDiff. The real loss we used here is and end to end diffusion loss?
A: Yes, you are right.  And the real loss is diffusion loss, the loss of AR Transformer is given while the backpropagation, which is similar with MAR.

Q: Besides, ‘x_n denotes the n-th known input latent’ means n-th latent feature which is pixel or latent level shaped as (h’,w’,c), not a token level shaped (1,1,c), no?
A:  x_n is a tensor  R^{bsz, h, w, c}  which is encoder by VAE model.  In AR Transformer,  x_n is patched and resize to a tensor  R^{bsz, (h/f * w/f), c *f *f}, which f is the compression ratio:
https://github.com/TransDiff/TransDiff/blob/4a501e0c3046a3c44f571d9fad541405c0ffcf25/models/transdiff.py#L177

In 1StepAR, f is 1, and in MRAR, f is 2.

**Comment by lawrence-cj** [comment]
Great. That is perfeclty clear. Thanks so much for your patience in answering.
